keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -230,80 +263,110 @@ scenario of wrong-single-stack pods...
 
 ### Service and EndpointSlice Reconciling
 
-If the dual-stack apiserver changes were to take effect in a cluster
-all at once when the feature went to beta, then this could cause
-Service/EndpointSlice flapping in dual-stack clusters during the
-upgrade when the `DualStackAPIServer` feature gate becomes enabled,
-because the older apiservers would be trying to enforce a single-stack
-`kubernetes.default` service, while the newer apiservers would be
-trying to make it be dual-stack.
-
-(Note that the problem is only with `Service` and `EndpointSlice`, not
-with `Endpoints` because `Endpoints` is inherently single-stack, so
-both old and new apiservers will always agree on which IPs should be
-listed there.)
-
-To prevent this, we can make some of the dual-stack apiserver support
-code be active even when the feature gate is disabled; the feature
-gate will control whether an apiserver in a dual-stack cluster tries
-to advertise *itself* as dual-stack, but not whether it allows *other*
-apiservers to advertise themselves as dual-stack. Thus:
-
-  - API server does not know about `DualStackAPIServer` (eg,
-    Kubernetes 1.21, regardless of single/dual stack configuration):
+#### Current Behavior
+
+Current kube-apiservers attempt to force the `kubernetes` Service
+definition to match its expectations on startup, but do not attempt to
+reconcile it after that point. Current apiservers always expect the
+Service to be single-stack, so any time a current apiserver reconciles
+the `kubernetes` Service, it will set it to be single-stack.
+
+For Endpoints/EndpointSlice, the apiserver can run in two different
+"endpoint reconciler" modes: `master-count` (the old algorithm), and
+`lease` (the current default, and preferred mode).
+
+In `master-count` mode, each apiserver periodically checks the
+`kubernetes` `Endpoints` object, and:
+
+  - If its own IP is not listed there, it adds it.
+
+  - If the total number of IPs listed there is now greater than the
+    expected number of masters, it removes a randomly-selected IP
+    other than its own.
+
+(The changes to `Endpoints` are likewise synced to `EndpointSlice`.)
+
+This means that when an apiserver is replaced, it may take arbitrarily
+long for the `Endpoints` to settle on the correct values (since none
+of the remaining apiservers knows which IP is the one that needs to be
+removed), but once it reaches the correct values, it will stay correct
+until the set of apiservers changes again.
+
+In the `lease` mode, each apiserver periodically writes its IP to a
```

Comment:
Could you ever have the case where different apiservers write different families? (i.e. one v4, one v6)? I assume this would be the symptom of a very broken cluster, and thus not really worth worrying about - but still interesting to ponder.

Reply:
So in order for that to happen they'd have to have different service CIDRs too... I assume they make sure that doesn't happen?